src/X86Optimize.cpp

```diff
@@ -0,0 +1,633 @@
+#include "X86Optimize.h"
+
+#include "CSE.h"
+#include "FindIntrinsics.h"
+#include "IR.h"
+#include "IRMatch.h"
+#include "IRMutator.h"
+#include "IROperator.h"
+#include "InstructionSelector.h"
+#include "Simplify.h"
+
+namespace Halide {
+namespace Internal {
+
+#if defined(WITH_X86)
+
+namespace {
+
+// i32(i16_a)*i32(i16_b) +/- i32(i16_c)*i32(i16_d) can be done by
+// interleaving a, c, and b, d, and then using dot_product.
+bool should_use_dot_product(const Expr &a, const Expr &b, std::vector<Expr> &result) {
+    Type t = a.type();
+    internal_assert(b.type() == t) << a << " and " << b << " don't match types\n";
+
+    if (!(t.is_int() && t.bits() == 32 && t.lanes() >= 4)) {
+        return false;
+    }
+
+    const Call *ma = Call::as_intrinsic(a, {Call::widening_mul});
+    const Call *mb = Call::as_intrinsic(b, {Call::widening_mul});
+    // dot_product can't handle mixed type widening muls.
+    if (ma && ma->args[0].type() != ma->args[1].type()) {
+        return false;
+    }
+    if (mb && mb->args[0].type() != mb->args[1].type()) {
+        return false;
+    }
+    // If the operands are widening shifts, we might be able to treat these as
+    // multiplies.
+    const Call *sa = Call::as_intrinsic(a, {Call::widening_shift_left});
+    const Call *sb = Call::as_intrinsic(b, {Call::widening_shift_left});
+    if (sa && !is_const(sa->args[1])) {
+        sa = nullptr;
+    }
+    if (sb && !is_const(sb->args[1])) {
+        sb = nullptr;
+    }
+    if ((ma || sa) && (mb || sb)) {
+        Expr a0 = ma ? ma->args[0] : sa->args[0];
+        Expr a1 = ma ? ma->args[1] : lossless_cast(sa->args[0].type(), simplify(make_const(sa->type, 1) << sa->args[1]));
+        Expr b0 = mb ? mb->args[0] : sb->args[0];
+        Expr b1 = mb ? mb->args[1] : lossless_cast(sb->args[0].type(), simplify(make_const(sb->type, 1) << sb->args[1]));
+        if (a1.defined() && b1.defined()) {
+            std::vector<Expr> args = {a0, a1, b0, b1};
+            result.swap(args);
+            return true;
+        }
+    }
+    return false;
+}
+
+/** A top-down code optimizer that replaces Halide IR with VectorInstructions specific to x86. */
+class Optimize_X86 : public InstructionSelector {
+public:
+    /** Create an x86 code optimizer. Processor features can be
+     * enabled using the appropriate flags in the target struct. */
+    Optimize_X86(const Target &target, const CodeGen_LLVM *codegen)
+        : InstructionSelector(target, codegen) {
+    }
+
+    using IRGraphMutator::mutate;
+    Expr mutate(const Expr &e) override {
+        Expr expr = IRGraphMutator::mutate(e);
+        internal_assert(expr.type() == e.type()) << "(X86Optimize) Found type mismatch: " << e << " -> " << expr << "\n";
+        return expr;
+    }
+
+protected:
+    bool should_peephole_optimize(const Type &type) {
```

Comment:
Not convinced this is necessary or helpful. It may well be that we want to apply these rewrites to scalars too.